<a href="https://colab.research.google.com/github/the1975/python_learning/blob/main/Data_Cleaning_Task_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task 5 Data Cleaning dan Data Transformation

### Nama : Leonardo yanuarianto
### NIM : 2513020074



## Tujuan Praktikum
Mahasiswa Mampu
1. Mengidentifikasi berbagai masalah data kompleks
2. Menentukan strategi cleaning yang tepat (tidak sekadar ikut template)
3. Melakukan transformasi data untuk kebutuhan analisis
4. Menyusun insight berbasis hasil preprocessing

lakukan Pre Processing data dari dataset dibawah ini
- **Netflix Movies and TV Shows Dataset** - https://www.kaggle.com/datasets/shivamb/netflix-shows



## Data Understanding

Dengan Teknik EDA tampilkan informasi dari masing-masing dataset
1. 5 data teratas
2. berapa jumlah baris dan kolom
3. type data pada masing-masing kolom

In [ ]:
# Import library yang dibutuhkan

import pandas as pd
import numpy as np
import kagglehub
import os

# Download latest version
path = kagglehub.dataset_download("shivamb/netflix-shows")

# Memunculkan isi direktori lokal
print(os.listdir(path))
print("Path to dataset files:", path)

Using Colab cache for faster access to the 'netflix-shows' dataset.
['netflix_titles.csv']
Path to dataset files: /kaggle/input/netflix-shows


In [ ]:
# Load dataset
df = pd.read_csv(path + '/netflix_titles.csv')
df.head()

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,NaN,United States,"September 25, 2021",2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm..."
1,s2,TV Show,Blood & Water,NaN,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t..."
2,s3,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",NaN,"September 24, 2021",2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...
3,s4,TV Show,Jailbirds New Orleans,NaN,NaN,NaN,"September 24, 2021",2021,TV-MA,1 Season,"Docuseries, Reality TV","Feuds, flirtations and toilet talk go down amo..."
4,s5,TV Show,Kota Factory,NaN,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...


In [ ]:
# Mengecek jumlah baris dan kolom
df.shape

(8807, 12)

In [ ]:
# Cek typetype data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


Analisis dari tiap dataset
- Berapa jumlah missing value?
- Kolom mana paling bermasalah?
- Tipe data sudah sesuai atau belum?

In [ ]:
# CekCek jumlah missing value
df.isnull().sum()

,0
show_id,0
type,0
title,0
director,2634
cast,825
country,831
date_added,10
release_year,0
rating,4
duration,3


In [ ]:
# Cek kolom mana yang paling bermasalah
df.isnull().sum().sort_values(ascending=False)

,0
director,2634
country,831
cast,825
date_added,10
rating,4
duration,3
show_id,0
type,0
title,0
release_year,0


In [ ]:
# Mengecek tipe data
df.dtypes

,0
show_id,object
type,object
title,object
director,object
cast,object
country,object
date_added,object
release_year,int64
rating,object
duration,object


Jawaban disini


1. Terdapat beberapa mising value pada Dataset ini, terutama pada kolom seperti:

   * `director` (2634)  
   * `cast` (825)
   * `country` (831)
   * `date_added` (10)
   * `rating` (4)
   * `duration` (3)
   
   Jadi jumlah missing value nya sebanyak (4307)

2. Kolom yang paling bermasalah adalah `director`, karena memiliki jumlah missing value terbanyak, Kolom `director` memiliki missing value terbanyak karena tidak semua konten di Netflix mencantumkan sutradara. Hal ini bisa terjadi karena beberapa jenis konten seperti acara TV, reality show, atau produksi tertentu tidak memiliki satu sutradara utama atau datanya tidak tersedia.

3. Berdasarkan `df.info()`, sebagian besar tipe data sudah sesuai, namun terdapat kolom seperti `date_added` yang bertipe object dan seharusnya diubah menjadi `datetime`. dengan mengubahnya menjadi `datetime`, data dapat diolah lebih lanjut seperti mengambil tahun, bulan, serta melakukan analisis tren penambahan konten dari waktu ke waktu.



## Data Cleaning
1 . Cari Missing Value masing-masing dataset

Tentukan strategi untuk tiap kolom pada masing-masing dataset:

- Hapus?
- Isi?
- Biarkan?

In [ ]:
# Strategi per kolom
# biarkan
df['director'].fillna('Unknown', inplace=True)
df['cast'].fillna('Unknown', inplace=True)
df['country'].fillna('Unknown', inplace=True)
df['duration'].fillna('Unknown', inplace=True)

# hapus
df.dropna(subset=['date_added'], inplace=True)
df = df.dropna(subset=['date_added'])

# isi
df['rating'].fillna('df[rating].mode()[0]', inplace=True)

/tmp/ipykernel_2939/2969426973.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['director'].fillna('Unknown', inplace=True)
/tmp/ipykernel_2939/2969426973.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try us

In [ ]:
# Cek hasilnya
df['date_added'].head()

print(df.isnull().sum())

show_id         0
type            0
title           0
director        0
cast            0
country         0
date_added      0
release_year    0
rating          0
duration        0
listed_in       0
description     0
dtype: int64


Jawaban disini

Strategi penanganan missing value pada dataset dilakukan berdasarkan tingkat kepentingan dan jumlah data yang hilang.
* Kolom seperti `director`, `cast`, dan `country` memiliki banyak missing value, sehingga tidak dihapus, melainkan diisi dengan nilai “Unknown” agar tidak kehilangan banyak data.

* Kolom `date_added` memiliki missing value yang sedikit dan penting untuk analisis waktu, sehingga baris yang mengandung missing value dihapus.
* Kolom `rating` diisi menggunakan nilai modus karena jumlah missing sedikit dan masih dapat diperkirakan.
* Kolom penting seperti `title`, `type`, dan `release_year` tidak boleh memiliki missing value, sehingga jika terdapat nilai kosong maka baris tersebut dihapus.

2. Cleaning Kolom Tanggal

In [ ]:
# date_added ubah ke datetime
df['date_added'] = pd.to_datetime(df['date_added'], errors='coerce')

df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8797 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   show_id       8797 non-null   object        
 1   type          8797 non-null   object        
 2   title         8797 non-null   object        
 3   director      8797 non-null   object        
 4   cast          8797 non-null   object        
 5   country       8797 non-null   object        
 6   date_added    8709 non-null   datetime64[ns]
 7   release_year  8797 non-null   int64         
 8   rating        8797 non-null   object        
 9   duration      8797 non-null   object        
 10  listed_in     8797 non-null   object        
 11  description   8797 non-null   object        
dtypes: datetime64[ns](1), int64(1), object(10)
memory usage: 893.4+ KB


3. Cek Duplikasi Data

In [ ]:
# Cek jumlah duplikat
df.duplicated().sum()

np.int64(0)

Jelaskan apakah duplikasi data perlu dihapus

jawab disini

- Tidak ada data duplikat yang harus dihapus.

## Data Transformation
1. buatlah feature engineering dengan membuat kolom baru dengan nama **year_added** dari kolom **date_added**

In [ ]:
# Menambahkan tahun. bulan, hari
df['year_added'] = df['date_added'].dt.year
df['month_added'] = df['date_added'].dt.month
df['day_added'] = df['date_added'].dt.day

df.head()

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description,year_added,month_added,day_added,kategori_konten
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,Unknown,United States,2021-09-25,2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm...",2021.0,9.0,25.0,NEW
1,s2,TV Show,Blood & Water,Unknown,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,2021-09-24,2021,TV-MA,2 Seasons,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t...",2021.0,9.0,24.0,NEW
2,s3,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",Unknown,2021-09-24,2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...,2021.0,9.0,24.0,NEW
3,s4,TV Show,Jailbirds New Orleans,Unknown,Unknown,Unknown,2021-09-24,2021,TV-MA,1 Season,"Docuseries, Reality TV","Feuds, flirtations and toilet talk go down amo...",2021.0,9.0,24.0,NEW
4,s5,TV Show,Kota Factory,Unknown,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,2021-09-24,2021,TV-MA,2 Seasons,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...,2021.0,9.0,24.0,NEW


2. Buat Kolom baru atau feature engineering dengan nama kolom **kategori_konten** dari kolom **release_year** sesuai dengan kategori umur konten dibawah ini
```
jika tahun kurang dari 2000 kategori OLD, selain itu jika tahun kurang dari 2015 MIDDLE selain itu NEW
```



In [ ]:
# Membuat function untuk menentukan kategori konten berdasarkan tahun rilis (release_year)
def kategori_konten(year):
    if year < 2000:
        return 'OLD'
    elif year < 2015:
        return 'MIDDLE'
    else:
        return 'NEW'

# Menambahkan kolom kategori_konten dari kolom relase_year
df['kategori_konten'] = df['release_year'].apply(kategori_konten)

# Cek hasil
df[['release_year', 'kategori_konten']].head()

,release_year,kategori_konten
0,2020,NEW
1,2021,NEW
2,2021,NEW
3,2021,NEW
4,2021,NEW


3. Encoding Kolom **Rating / Type**

In [ ]:
# Encoding kolom rating/type
df_encoded = pd.get_dummies(df, columns=['rating', 'type'], prefix=['rating', 'type'])

# Cek hasilnya
df_encoded.head()

,show_id,title,director,cast,country,date_added,release_year,duration,listed_in,description,...,rating_TV-G,rating_TV-MA,rating_TV-PG,rating_TV-Y,rating_TV-Y7,rating_TV-Y7-FV,rating_UR,rating_df[rating].mode()[0],type_Movie,type_TV Show
0,s1,Dick Johnson Is Dead,Kirsten Johnson,Unknown,United States,2021-09-25,2020,90 min,Documentaries,"As her father nears the end of his life, filmm...",...,False,False,False,False,False,False,False,False,True,False
1,s2,Blood & Water,Unknown,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,2021-09-24,2021,2 Seasons,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t...",...,False,True,False,False,False,False,False,False,False,True
2,s3,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",Unknown,2021-09-24,2021,1 Season,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...,...,False,True,False,False,False,False,False,False,False,True
3,s4,Jailbirds New Orleans,Unknown,Unknown,Unknown,2021-09-24,2021,1 Season,"Docuseries, Reality TV","Feuds, flirtations and toilet talk go down amo...",...,False,True,False,False,False,False,False,False,False,True
4,s5,Kota Factory,Unknown,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,2021-09-24,2021,2 Seasons,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...,...,False,True,False,False,False,False,False,False,False,True


4. lakukan filter data untuk menampilkan data yang bersalah dari country United State

In [ ]:
# Filter data country US
df_us = df[df['country'].str.contains('United States', na=False)]

# Cek hasilnya
df_us.head()

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description,year_added,month_added,day_added,kategori_konten
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,Unknown,United States,2021-09-25,2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm...",2021.0,9.0,25.0,NEW
7,s8,Movie,Sankofa,Haile Gerima,"Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...","United States, Ghana, Burkina Faso, United Kin...",2021-09-24,1993,TV-MA,125 min,"Dramas, Independent Movies, International Movies","On a photo shoot in Ghana, an American model s...",2021.0,9.0,24.0,OLD
9,s10,Movie,The Starling,Theodore Melfi,"Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...",United States,2021-09-24,2021,PG-13,104 min,"Comedies, Dramas",A woman adjusting to life after a loss contend...,2021.0,9.0,24.0,NEW
15,s16,TV Show,Dear White People,Unknown,"Logan Browning, Brandon P. Bell, DeRon Horton,...",United States,2021-09-22,2021,TV-MA,4 Seasons,"TV Comedies, TV Dramas",Students of color navigate the daily slights a...,2021.0,9.0,22.0,NEW
27,s28,Movie,Grown Ups,Dennis Dugan,"Adam Sandler, Kevin James, Chris Rock, David S...",United States,2021-09-20,2010,PG-13,103 min,Comedies,Mourning the loss of their beloved junior high...,2021.0,9.0,20.0,MIDDLE


## Analisi dan Penalaran
mahasiswa harus menjawab
1. Cleaning Decision
- Kenapa missing value tidak semuanya dihapus?
- Kolom mana yang paling “berisiko”?

jawab disini

1. * Missing value tidak semuanya dihapus karena jika terlalu banyak data yang dihapus, maka akan menyebabkan kehilangan informasi yang signifikan dan membuat dataset menjadi kurang representatif. Oleh karena itu, beberapa kolom yang memiliki banyak missing value lebih baik diisi dengan nilai tertentu seperti “Unknown” daripada dihapus.

   * Kolom yang paling berisiko adalah `director`, karena memiliki jumlah missing value paling banyak. Jika kolom ini dihapus secara langsung, maka akan menghilangkan banyak baris data seperti keterangan diatas. Selain itu, kolom seperti `cast` dan `country` juga memiliki risiko karena cukup banyak data yang kosong.

2. Insight Dataset
Minimal 5 insight:

- Negara dengan konten terbanyak
- Tahun dengan rilis terbanyak
- Distribusi kategori konten
- Pola data kosong

jawab disini

2.  * **Negara dengan konten terbanyak**

     Negara dengan jumlah konten terbanyak adalah United States, yang menunjukkan dominasi industri hiburan Amerika dalam platform Netflix.

    * **Tahun dengan rilis terbanyak**

     Tahun dengan rilis terbanyak berada di sekitar 2018–2020, menunjukkan peningkatan produksi dan distribusi konten secara signifikan pada periode tersebut.

    * **Distribusi kategori konten**

     Sebagian besar konten termasuk dalam kategori NEW, yang berarti mayoritas konten yang tersedia adalah produksi setelah tahun 2015.

    * **Pola data kosong**

     Missing value banyak ditemukan pada kolom seperti director, cast, dan country, yang menunjukkan bahwa informasi terkait produksi tidak selalu lengkap.

    * **Dominasi tipe konten**

     Konten dengan tipe Movie lebih banyak dibandingkan TV Show, yang menunjukkan bahwa film masih menjadi jenis konten utama di dataset ini.

3. Dampak Data Cleaning
- Apa perubahan sebelum vs sesudah cleaning?
- Apakah data lebih siap untuk analisis?

jawab disini

3. * Sebelum cleaning, dataset memiliki banyak missing value, tipe data yang belum sesuai, serta kemungkinan data yang tidak konsisten. Setelah cleaning, missing value telah ditangani, tipe data seperti `date_added` sudah diubah menjadi `datetime`, dan data menjadi lebih terstruktur serta konsisten.

   * Ya, setelah proses cleaning, data menjadi lebih siap untuk dianalisis karena sudah bebas dari masalah utama seperti missing value yang tidak tertangani, tipe data yang salah, dan potensi kesalahan dalam pengolahan data. Hal ini memungkinkan analisis yang lebih akurat dan valid.